src/deriver/ScopedKeys.js

```diff
@@ -0,0 +1,111 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const HKDF = require('node-hkdf');
+const base64url = require('base64url');
+
+const KEY_LENGTH = 32;
+
+/**
+ * Scoped key deriver
+ * @desc Used by the Firefox Accounts content server
+ * @module deriver-ScopedKeys
+ * @example
+ * ```js
+ * const scopedKeys = new fxaCryptoDeriver.ScopedKeys();
+ *
+ * return scopedKeys.deriveScopedKey({
+ *   identifier: 'https://identity.mozilla.com/apps/notes',
+ *   inputKey: 'bc3851e9e610f631df94d7883d5defd5e5f55ab520bd5a9ae33dae26575c6b1a',
+ *   keyMaterial: '0000000000000000000000000000000000000000000000000000000000000000',
+ *   timestamp: 1494446722583
+ * });
+ * ```
+ */
+class ScopedKeys {
+  /**
+   * Derive a scoped key
+   * @method deriveScopedKey
+   * @param {object} options - required set of options to derive a scoped key
+   * @param {string} options.inputKey - input key hex string that the scoped key is derived from
+   * @param {string} options.keyMaterial - a 32-byte hex string of additional entropy specific to this scoped key
+   * @param {number} options.timestamp
+   *   A 13-digit number, the timestamp in milliseconds at which this scoped key most recently changed
+   * @param {string} options.identifier - a unique URI string identifying the requested scoped key
+   * @returns {Promise}
+   */
+  deriveScopedKey(options) {
+    return new Promise((resolve) => {
+      if (! options.inputKey) {
+        throw new Error('inputKey required');
+      }
+
+      if (! options.keyMaterial) {
+        throw new Error('keyMaterial required');
+      }
+
+      if (! options.timestamp) {
+        throw new Error('timestamp required');
+      }
+
+      if (! options.identifier) {
+        throw new Error('identifier required');
+      }
+
+      if (options.timestamp.toString().length !== 13) {
+        throw new Error('timestamp must be a 13-digit number');
+      }
+
+      const context = 'identity.mozilla.com/picl/v1/scoped_key\n' +
+        options.identifier;
+      const contextKid = 'identity.mozilla.com/picl/v1/scoped_kid\n' +
+        options.identifier;
+      const contextHex = Buffer.from(context).toString('hex');
+      const contextKidHex = Buffer.from(contextKid).toString('hex');
+      const scopedKey = {
+        kty: 'oct',
+        scope: options.identifier,
+      };
+
+      this._deriveHKDF(options.keyMaterial, options.inputKey, contextHex, KEY_LENGTH)
+        .then((key) => {
+          scopedKey.k = base64url(key);
```

Comment:
As far as I know, this `base64url(key)` operation is key-dependant, and can take a different amount of time depending on the specific contents of the key.  I'm not sure what (if anything) to do about this, because I haven't been able to find a time-invariant JS implementation of this.

Reply:
from mtg: we are not using btoa(), so there is that.